File: #1498.number-of-subsequences-that-satisfy-the-given-sum-condition.cpp
class Solution {
public:
    int numSubseq(vector<int>& nums, int target) {

        int size = nums.size();
        int mod = 1e9+7;
        vector<int> v(size, 1);
        for (int i = 1; i < size; ++i) {
            v[i] = (v[i-1] << 1) % mod;
        }
        
        std::sort(nums.begin(), nums.end());
        int result = 0;
        for (int l = 0, r = size - 1; l <= r;) {
            if (nums[l] + nums[r] > target) {
                --r;
            } else {
                result = (result + v[r - l]) % mod;
                ++l;
            }
        }
        return result;   
    }
};

File: #1498.number-of-subsequences-that-satisfy-the-given-sum-condition.ts
function numSubseq(nums: number[], target: number): number {
    nums.sort((a, b) => a - b);
    
    let res = 0;
    const mod = Math.pow(10, 9) + 7;
    let [left, right] = [0, nums.length - 1];
    const pows: number[] = [1];
    for (let i = 1; i < nums.length; i++) {
        pows.push(pows[i - 1] * 2 % mod);
    }

    while (left <= right) {
        if (nums[left] + nums[right] <= target) res = (res + pows[right - left++]) % mod;
        else right--;
    }
    return res;
};
